deepchem/splits/__init__.py

```diff
@@ -5,188 +5,291 @@
-        frac_valid=frac_valid, log_every_n=log_every_n)
-    train_dataset = dataset.select(train_dir, train_inds)
-    if valid_dir is not None:
-      valid_dataset = dataset.select(valid_dir, valid_inds)
-    else:
-      valid_dataset = None
-    test_dataset = dataset.select(test_dir, test_inds)
-
-    return train_dataset, valid_dataset, test_dataset
-
-  def train_test_split(self, samples, train_dir, test_dir, seed=None,
-                       frac_train=.8):
-    """
+        log("Computing train/valid/test indices", self.verbosity)
+        train_inds, valid_inds, test_inds = self.split(
+            dataset,
+            frac_train=frac_train, frac_test=frac_test,
+            frac_valid=frac_valid, log_every_n=log_every_n)
+        train_dataset = dataset.select(train_dir, train_inds)
+        if valid_dir is not None:
+            valid_dataset = dataset.select(valid_dir, valid_inds)
+        else:
+            valid_dataset = None
+        test_dataset = dataset.select(test_dir, test_inds)
+
+        return train_dataset, valid_dataset, test_dataset
+
+    def train_test_split(self, samples, train_dir, test_dir, seed=None,
+                         frac_train=.8):
+        """
     Splits self into train/test sets.
 
     Returns Dataset objects.
     """
-    valid_dir = None
-    train_samples, _, test_samples = self.train_valid_test_split(
-        samples, train_dir, valid_dir, test_dir,
-        frac_train=frac_train, frac_test=1-frac_train, frac_valid=0.)
-    return train_samples, test_samples
-
-  def split(self, samples, frac_train=None, frac_valid=None, frac_test=None,
-            log_every_n=None):
-    """
+        valid_dir = None
+        train_samples, _, test_samples = self.train_valid_test_split(
+            samples, train_dir, valid_dir, test_dir,
+            frac_train=frac_train, frac_test=1 - frac_train, frac_valid=0.)
+        return train_samples, test_samples
+
+    def split(self, samples, frac_train=None, frac_valid=None, frac_test=None,
+              log_every_n=None):
+        """
     Stub to be filled in by child classes.
     """
-    raise NotImplementedError
+        raise NotImplementedError
 
-class MolecularWeightSplitter(Splitter):
+class StratifiedSplitter(Splitter):
+    """
+  Class for doing stratified splits -- where data is too sparse to do regular splits
   """
+
+    def __randomize_arrays(self, array_list):
+      #assumes that every array is of the same dimension
+      numRows = array_list[0].shape[0]
+      perm = np.random.permutation(numRows)
+      for array in array_list:
+          array = array[perm]
+      return array_list
+
+    def __generate_required_hits(self, w, frac_split):
+      required_hits = (w != 0).sum(0)#returns list of per column sum of non zero elements
+      for colHits in required_hits:
+        colHits = int(frac_split * colHits)
+      return required_hits
+
+    def __generate_required_index(self, w, required_hit_list):
+      colIndex = 0
```

Comment:
Can you swap camel-case variables like `colIndex` to use underscores? `col_index` in this case.